sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/WALUtils.scala

```diff
@@ -0,0 +1,281 @@
+    output.close()
+  }
+
+  def updateFromDeltaFile(
+      fm: CheckpointFileManager,
+      fileToRead: Path,
+      keySchema: StructType,
+      valueSchema: StructType,
+      newRocksDb: OptimisticTransactionDbInstance,
+      sparkConf: SparkConf): Unit = {
+    var input: DataInputStream = null
+    val sourceStream = try {
+      fm.open(fileToRead)
+    } catch {
+      case f: FileNotFoundException =>
+        throw new IllegalStateException(
+          s"Error reading delta file $fileToRead of $this: $fileToRead does not exist",
+          f)
+    }
+    try {
+      input = decompressStream(sourceStream, sparkConf)
+      var eof = false
+
+      while (!eof) {
+        val keySize = input.readInt()
+        if (keySize == EOF_MARKER) {
+          eof = true
+        } else if (keySize < 0) {
+          newRocksDb.abort
+          newRocksDb.close()
+          throw new IOException(
+            s"Error reading delta file $fileToRead of $this: key size cannot be $keySize")
+        } else {
+          val keyRowBuffer = new Array[Byte](keySize)
+          ByteStreams.readFully(input, keyRowBuffer, 0, keySize)
+
+          val keyRow = new UnsafeRow(keySchema.fields.length)
+          keyRow.pointTo(keyRowBuffer, keySize)
+
+          val valueSize = input.readInt()
+          if (valueSize < 0) {
+            newRocksDb.remove(key = keyRow)
+          } else {
+            val valueRowBuffer = new Array[Byte](valueSize)
+            ByteStreams.readFully(input, valueRowBuffer, 0, valueSize)
+            val valueRow = new UnsafeRow(valueSchema.fields.length)
+            // If valueSize in existing file is not multiple of 8, floor it to multiple of 8.
+            // This is a workaround for the following:
+            // Prior to Spark 2.3 mistakenly append 4 bytes to the value row in
+            // `RowBasedKeyValueBatch`, which gets persisted into the checkpoint data
+            valueRow.pointTo(valueRowBuffer, (valueSize / 8) * 8)
+            newRocksDb.put(keyRow, valueRow)
+          }
+        }
+      }
+    } finally {
+      if (input != null) input.close()
+    }
+  }
+
+  /*
+   * Try to cancel the underlying stream and safely close the compressed stream.
+   *
+   * @param compressedStream the compressed stream.
+   * @param rawStream the underlying stream which needs to be cancelled.
+   */
+  def cancelDeltaFile(
+      compressedStream: DataOutputStream,
+      rawStream: CancellableFSDataOutputStream): Unit = {
+    try {
+      if (rawStream != null) rawStream.cancel()
+      IOUtils.closeQuietly(compressedStream)
+    } catch {
+      case e: FSError if e.getCause.isInstanceOf[IOException] =>
+      // Closing the compressedStream causes the stream to write/flush flush data into the
+      // rawStream. Since the rawStream is already closed, there may be errors.
+      // Usually its an IOException. However, Hadoop's RawLocalFileSystem wraps
+      // IOException into FSError.
+    }
+  }
+
+  def uploadFile(
```

Comment:
`uploadFile` and `downloadFile` are new util function used by RocksDBStateStore.